### ISSUE TITLE
RFE: BPF manager/daemon

### ISSUE THREAD

**Issue by anitazha** [systemd/systemd#26510]
### Component

other

### Is your feature request related to a problem? Please describe

_We've discussed a lot of this offline so making the issue to help contain discussion._

@anakryiko has been collecting uses cases/requirements that a BPF manager/daemon could fulfill:
- Unprivileged applications that need to use BPF (without coordination). The daemon could perform a permissions check against an allow list to avoid giving out root/CAP_BPF to every application needs to use BPF. There could also be additional safety checks to do thinks like checking the  signature of bpf programs, sanitizing return values, and other restrictions that are relevant to the next point:
- Applications requiring coordination. If there are multiple applications that want to work with a given kernel object (e.g., network interface), there needs to be some sort of coordination between developers to make sure their BPF programs can co-exist. For example the kernel allows just one XDP program per each network interface. We currently do this internally with different chainer implementations being built in different teams to coordinate programs. A centralized BPF daemon could be the coordinator instead.

The daemon does not intend to replace instances where the application would already need root for other operations, and the BPF program does not need coordination.

There are some other use cases that could be mutually beneficial internally and upstream, but are not the primary focus of this RFE. For example, having a place to put BPF scheduler policies for [sched_ext](https://lore.kernel.org/lkml/20221130082313.3241517-29-tj@kernel.org/) and performing BPF monitoring and observation on a host.

### Describe the solution you'd like

A BPF daemon that would do the above. Given that we're still in the discussion phase there are no promises about when this will see the light. Especially since we are looking at prototyping this internally first before upstreaming.

In some discussions with maintainers and Alexei Starovoitov there was consensus around building the daemon and using varlink as the communication interface. I guess my only concern here is that it runs counter to what we do with daemons currently: we use varlink for the internal interface and dbus as the external interface.

There was another concern with circular dependency since systemd currently uses BPF and we would need to coordinate with the BPF daemon. Seems fixable.

### Describe alternatives you've considered

At one point we discussed making this a part of PID 1/systemd, without another daemon. But I think given the current requirements, having a dedicated BPF daemon/ecosystem makes sense here.

cc @anakryiko @davide125 

### The systemd version you checked that didn't have the feature you are asking for

252

**Comment by lperkov** [comment]
Hello @anitazha

Happy to see your post here, we have been discussing the same feature in parallel with the systemd maintainers.

In most of the points I agree with you but we do think the code should really be part of the systemd ecosystem as it will then be shipped by default on majority of distributions which is a very important aspect of having de-facto standard BPF daemon. Otherwise one can choose to extend any other BPF loader out there and there. There are already many and it makes no sense to go into this direction as this is what already exist today and does not serve the purpose.

Looking at the eBPF program lifecycle when managed within the OS it is really similar to the daemon management:

* loading the eBPF program (similar to starting a daemon/service)
* unloading the eBPF program (similar to stoping a daemon/service)

As part of the requirements, one should be able to access eBPF maps via dbus API so any consumer can read and write values into them. The map "description" is something that should be part of the configuration file and it is up to the eBPF consumer to provide the map description (which could be also provided by the eBPF program developer to begin with).

We plan to develop and submit this exciting feature upstream to systemd and it would be interesting to collaborate on this activity. Developing this internally within the organization without wider collaboration does not benefit the ecosystem as a whole.

We are planning to open a first pull request during Q2, unless there are major objections.

cc @dpintaric @lukakudra

**Comment by bbanelli** [comment]
Historically as a good rule of thumb, if you're looking to add massive contributions to an open source project, it's important to start small and incrementally. 

As far as I see it, first step that makes most sense would simply be to add infrastructure to load and unload eBPF program. Once that is in place it can be extended to incrementally add support for other requirements as listed above.

As discussed on FOSDEM with @poettering @bluca, do you think this is a good approach to get started with eBPF integration within systemd?

**Comment by poettering** [comment]
> In some discussions with maintainers and Alexei Starovoitov there was consensus around building the daemon and using varlink as the communication interface. I guess my only concern here is that it runs counter to what we do with daemons currently: we use varlink for the internal interface and dbus as the external interface.

That's not entirely true anymore. the "userdb" logic makes varlink public, see here: https://systemd.io/USER_GROUP_API/

**Comment by poettering** [comment]
> There was another concern with circular dependency since systemd currently uses BPF and we would need to coordinate with the BPF daemon. Seems fixable.

Yeah, key is that PID 1 never talks to bpf daemon directly. Instead it probably should always fork something of that talks to bpfd, for example the code that we fork off for every process we start that eventually does execve(). That's where we currently do NSS (i.e. user name lookups and such), which is a similar case of IPC.

**Comment by poettering** [comment]
one stumbling block I see: i figure people will eventually use polkit auth for bpf stuff. polkit is a dbus thing though. which creates two problems:

1. it currently authenticates dbus peers, not arbitrary socket peers. 
2. polkit (and dbus) are late boot things, but we probably want bpfd to work during early boot already.

I think these are addressable:

1. if need be polkit can be extended to authenticate pidfds or similar. we'd then derive a pidfd of our varlink socket peers, and then go via dbus to authenticate that.
2. we need to be careful to implement a safe fallback if polkit/dbus is not up yet, i.e. default to a simple policy of "root can use bpfd but noone else" until dbus is up. We already have apis for creating bus connections "passively" that remain dormant until dbus actually starts up. so the idea would be to ask polkit once polkit is available, and just not bother with polkit before.

that all said, initial versions probably should not bother with polkit at all.

**Comment by poettering** [comment]
but anyway, all in all i am onboard.

**Comment by lperkov** [comment]
@poettering, @bluca, @bbanelli, @anitazha just to share that we started the activity in this branch:

https://github.com/sartura/systemd/commits/feature-ebpf

It is still work in progress and we will rebase it often until we make the pull request.

**Comment by kanashimia** [comment]
Just noting that it seems to be some similar effort occurring at https://github.com/redhat-et/bpfd

**Comment by astoycos** [comment]
Hi @anitazha @lperkov @poettering et all 👋  I'm a maintainer for [bpfd](https://bpfd.netlify.app/) and we do in fact already cover the main two use cases listed above, and have instructions for [deploying bpfd as a systemd service](https://bpfd.netlify.app/getting-started/tutorial/#systemd-service). 

We'd be happy to work with any new folks who may have suggestions on new use cases or improvements! Please reach out to us in Kubernetes slack at [#bpfd](https://kubernetes.slack.com/archives/C04UJBW2553) or feel free to open issues/ ask more questions here 👍

**Comment by DaanDeMeyer** [comment]
As an update to this discussion, for the unprivileged part of bpfd we're proposing an in-kernel solution; https://lore.kernel.org/bpf/CAEiveUeDLr00SjyU=SMSc4XbHSA6LTn4U2DHr12760rbo5WqSw@mail.gmail.com/T/

Assuming that is accepted the unprivileged BPF part of this proposal is not applicable anymore and we'll be focused primarily on monitoring bpf and providing data on what bpf programs are running in the system.

**Comment by lperkov** [comment]
@DaanDeMeyer, thanks for sharing this here. Do you have any inputs on what would be best way to read the data from the bpf-maps and expose them via D-Bus?
